File: src/adhush/detect/jingle.py
"""Break jingles (ADR 0020, after AdVent): the sting a channel plays into and
out of every break, learned rather than hard-coded.

Every break the engine ends contributes the seconds around its start (an
opener) and its end (a closer) as *candidates*, cut from a rolling ring of
chroma blocks. A candidate that turns out to open several different breaks
is promoted to a jingle and starts voting: an opener heard live votes 1.0
for ``hold_s`` (it carries the logo weight, so it mutes alone); a closer
heard while muted is program evidence (``program_present``), which ends a
break as fast as a returning logo. "Not an ad" after a jingle-driven mute
counts against that jingle; enough wrong calls demote it to a candidate.
Same TSV as the phone (``jingles.tsv``), so a memory can move.

Families and hours (ADR 0027): a channel re-cuts its sting per show, so a
sting is also recognised transposed by a few semitones (the 12 chroma bits
rotate) or played a little faster or slower (the stored blocks stretch), at a
higher bar; a variant hit counts for the jingle it resembles, so one memory
covers the family. Every jingle also remembers the local hours it opened
breaks in: heard again at one of those hours it is trusted a break earlier
and matched a little more loosely; a replay, which has no wall clock, learns
no hours.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import ClassVar

import numpy as np
import numpy.typing as npt

from adhush.config import JingleConfig
from adhush.detect.base import Detector
from adhush.events import AudioEvent, DetectorVote
from adhush.fingerprint.audio_chroma import CHROMA_BITS, agreement, chroma_bits

_MASK = (1 << CHROMA_BITS) - 1


def pitch_rotate(bits: int, semitones: int) -> int:
    """The chroma word of the same block transposed up by ``semitones``
    (down when negative): pitch class c moves to c + k, and class c lives in
    bit ``11 - c``, so up is a rotate towards bit 0."""
    k = semitones % CHROMA_BITS
    if k == 0:
        return bits & _MASK
    return ((bits >> k) | (bits << (CHROMA_BITS - k))) & _MASK


def tempo_stretch(blocks: list[int], factor: float) -> list[int]:
    """The same sting played at ``1 / factor`` speed: ``factor`` 1.1 is 10 % slower."""
    n = max(1, round(len(blocks) * factor))
    return [blocks[min(len(blocks) - 1, int(i / factor))] for i in range(n)]


def local_hour(wall: float) -> int:
    return datetime.fromtimestamp(wall).hour  # noqa: DTZ006 — the viewer's own clock, by design


@dataclass
class Jingle:
    id: int
    kind: str  # "open" | "close"
    blocks: list[int]
    hits: int
    false_hits: int
    created_ts: float
    hours: set[int] = field(default_factory=set)  # local hours this sting opened a break in
    last_wall: float = 0.0  # wall time of the last break that counted as a hit; 0 unknown


class JingleDetector(Detector):
    name: ClassVar[str] = "jingle"
    needs_audio: ClassVar[bool] = True

    def __init__(self, config: JingleConfig) -> None:
        self._cfg = config
        self._ring: list[tuple[float, int]] = []
        self._pending: list[npt.NDArray[np.float32]] = []
        self._pending_n = 0
        self._block_start: float | None = None
        self._jingles: list[Jingle] = []
        self._next_id = 1
        self._pending_close: float | None = None
        self._open_hit_ts: float | None = None
        self._open_hit_id: int | None = None
        self._close_hit_ts: float | None = None
        self._last_agreement = 0.0
        self._last_variant = ""
        self._last_ts = 0.0
        self._hour: int | None = None
        self._pending_close_hour: int | None = None
        self._pending_close_wall: float | None = None
        self._load()

    # -- sizes ---------------------------------------------------------------

    @property
    def _sting_blocks(self) -> int:
        return max(2, int(self._cfg.jingle_s / self._cfg.sample_interval_s))

    @property
    def _candidate_blocks(self) -> int:
        return max(self._sting_blocks, int(self._cfg.candidate_s / self._cfg.sample_interval_s))

    # -- persistence (j<TAB>id<TAB>kind<TAB>hits<TAB>false<TAB>created<TAB>b,b,b<TAB>h,h) --

    def _path(self) -> Path | None:
        return Path(self._cfg.file) if self._cfg.file else None

    def _load(self) -> None:
        path = self._path()
        if path is None or not path.is_file():
            return
        for line in path.read_text().splitlines():
            p = line.split("\t")
            if p[0] != "j" or len(p) < 7:
                continue
            blocks = [int(b) for b in p[6].split(",") if b]
            hours = {int(h) for h in p[7].split(",") if h} if len(p) > 7 else set()  # v1 lines carry none
            last_wall = float(p[8]) if len(p) > 8 and p[8] else 0.0
            self._jingles.append(Jingle(int(p[1]), p[2], blocks, int(p[3]), int(p[4]), float(p[5]), hours, last_wall))
            self._next_id = max(self._next_id, int(p[1]) + 1)

    def _save(self) -> None:
        path = self._path()
        if path is None:
            return
        path.parent.mkdir(parents=True, exist_ok=True)
        tmp = path.with_suffix(path.suffix + ".tmp")
        lines = ["# adhush jingles v2\tid\tkind\thits\tfalse_hits\tcreated\tblocks\thours\tlast_wall"]
        lines += [
            f"j\t{j.id}\t{j.kind}\t{j.hits}\t{j.false_hits}\t{j.created_ts}\t{','.join(map(str, j.blocks))}"
            f"\t{','.join(map(str, sorted(j.hours)))}\t{j.last_wall}"
            for j in self._jingles
        ]
        tmp.write_text("\n".join(lines) + "\n")
        tmp.replace(path)

    # -- audio ---------------------------------------------------------------

    def warmup(self) -> None:
        self._ring.clear()
        self._pending.clear()
        self._pending_n = 0
        self._block_start = None
        self._pending_close = None
        self._open_hit_ts = self._close_hit_ts = None

    def observe_audio(self, event: AudioEvent) -> None:
        start = self._block_start if self._block_start is not None else event.ts
        self._pending.append(event.samples)
        self._pending_n += len(event.samples)
        need = round(self._cfg.sample_interval_s * event.sample_rate)
        while self._pending_n >= need:
            all_samples = np.concatenate(self._pending)
            head, rest = all_samples[:need], all_samples[need:]
            self._ring.append((start, chroma_bits(head, event.sample_rate)))
            keep = int(self._cfg.history_s / self._cfg.sample_interval_s)
            if len(self._ring) > keep:
                del self._ring[: len(self._ring) - keep]
            start += need / event.sample_rate
            self._pending = [rest] if len(rest) else []
            self._pending_n = len(rest)
            self._on_sample(start)
        self._block_start = start

    def _blocks_between(self, t0: float, t1: float) -> list[int]:
        return [bits for ts, bits in self._ring if t0 <= ts < t1]

    def _best_agreement(self, live: list[int], candidate: list[int]) -> float:
        if len(live) < self._sting_blocks or len(candidate) < len(live):
            return 0.0
        return max(
            agreement(live, candidate[off : off + len(live)])
            for off in range(len(candidate) - len(live) + 1)
        )

    def _variants(self, blocks: list[int]) -> list[tuple[list[int], str]]:
        """The stored sting as heard, then every transposition and stretch the
        config allows, each labelled; the identity comes first and unlabelled."""
        out: list[tuple[list[int], str]] = [(blocks, "")]
        shifts = [k for k in range(-self._cfg.pitch_shifts, self._cfg.pitch_shifts + 1) if k]
        tol = self._cfg.tempo_tolerance
        factors = [1.0 - tol, 1.0 + tol] if tol > 0 else []
        for k in [0, *shifts]:
            rotated = [pitch_rotate(b, k) for b in blocks] if k else blocks
            for f in factors:
                out.append((tempo_stretch(rotated, f), f"{k:+d}st/{f:.2f}x"))
            if k:
                out.append((rotated, f"{k:+d}st"))
        return out

    def _best_family(self, live: list[int], jingle_blocks: list[int]) -> tuple[float, str]:
        """Best agreement of ``live`` against the sting or any variant of it, the
        variants docked ``family_penalty`` so the sting as heard always wins a tie."""
        best, label = 0.0, ""
        for blocks, variant in self._variants(jingle_blocks):
            a = self._best_agreement(live, blocks) - (self._cfg.family_penalty if variant else 0.0)
            if a > best:
                best, label = a, variant
        return best, label

    def _same_sting(self, a: list[int], b: list[int]) -> bool:
        core = self._sting_blocks
        return any(
            self._best_family(a[i : i + core], b)[0] >= self._cfg.min_agreement
            for i in range(len(a) - core + 1)
        )

    def _trusted(self, j: Jingle) -> bool:
        """Promoted outright, or one break short of it at an hour it has opened breaks in before."""
        if j.hits >= self._cfg.promote_hits:
            return True
        return self._hour is not None and self._hour in j.hours and j.hits >= self._cfg.promote_hits - 1 > 0

    def _threshold(self, j: Jingle) -> float:
        bonus = self._cfg.hour_bonus if self._hour is not None and self._hour in j.hours else 0.0
        return self._cfg.min_agreement - bonus

    def _on_sample(self, ts: float) -> None:
        self._last_ts = ts
        if self._pending_close is not None and ts >= self._pending_close + self._cfg.candidate_s - 2.0:
            end, self._pending_close = self._pending_close, None
            hour, self._pending_close_hour = self._pending_close_hour, None
            wall, self._pending_close_wall = self._pending_close_wall, None
            self._add_candidate("close", self._blocks_between(end - 2.0, end + self._cfg.candidate_s - 2.0), ts, hour, wall)
        live = [bits for _, bits in self._ring[-self._sting_blocks :]]
        if len(live) < self._sting_blocks:
            return
        best: Jingle | None = None
        best_a = 0.0
        best_v = ""
        for j in self._jingles:
            if not self._trusted(j):
                continue
            a, variant = self._best_family(live, j.blocks)
            if a >= self._threshold(j) and a > best_a:
                best, best_a, best_v = j, a, variant
        if best is None:
            return
        self._last_agreement, self._last_variant = best_a, best_v
        if best.kind == "open":
            if self._open_hit_ts is None or ts - self._open_hit_ts > self._cfg.hold_s:
                self._open_hit_ts, self._open_hit_id = ts, best.id
        else:
            self._close_hit_ts = ts

    def _add_candidate(
        self, kind: str, blocks: list[int], ts: float, hour: int | None = None, wall: float | None = None,
    ) -> None:
        if len(blocks) < self._sting_blocks:
            return
        for j in self._jingles:
            if j.kind == kind and self._same_sting(blocks, j.blocks):
                # Heard again within the spacing: the same session, not another break; it counts once.
                if wall is not None and j.last_wall > 0.0 and wall - j.last_wall < self._cfg.hit_spacing_s:
                    return
                j.hits += 1
                if hour is not None:
                    j.hours.add(hour)
                if wall is not None:
                    j.last_wall = wall
                self._save()
                return
        self._jingles.append(
            Jingle(self._next_id, kind, blocks, 1, 0, ts, {hour} if hour is not None else set(), wall or 0.0)
        )
        self._next_id += 1
        candidates = [j for j in self._jingles if j.hits < self._cfg.promote_hits]
        while len(candidates) > self._cfg.max_candidates:
            victim = min(candidates, key=lambda j: j.created_ts)
            self._jingles.remove(victim)
            candidates.remove(victim)
        self._save()

    # -- what the engine tells it ---------------------------------------------

    def tick(self, wall: float) -> None:
        self._hour = local_hour(wall)

    def learn_break(self, start_ts: float, end_ts: float, wall_start: float | None = None) -> None:
        """A real break ran from ``start_ts`` to ``end_ts`` (media time); it
        began at ``wall_start`` on the viewer's clock when the engine has one."""
        if end_ts - start_ts < self._cfg.min_break_s:
            return  # a mute that ended in seconds was no break (ADR 0027, amended)
        hour = local_hour(wall_start) if wall_start is not None else self._hour
        lead = self._cfg.candidate_s - 2.0
        self._add_candidate("open", self._blocks_between(start_ts - lead, start_ts + 2.0), end_ts, hour, wall_start)
        self._pending_close = end_ts
        self._pending_close_hour = hour
        self._pending_close_wall = wall_start

    def forget_all(self) -> None:
        """Drop every sting and candidate."""
        self._jingles.clear()
        self._open_hit_ts = self._open_hit_id = None
        self._close_hit_ts = self._pending_close = None
        self._save()

    def user_says_program(self, ts: float) -> None:
        if self._open_hit_id is not None and self._open_hit_ts is not None and ts - self._open_hit_ts < 60.0:
            for j in self._jingles:
                if j.id == self._open_hit_id:
                    j.false_hits += 1
                    if j.false_hits >= j.hits:
                        j.hits = 0
                    self._save()
        self._open_hit_ts = self._open_hit_id = None
        self._close_hit_ts = None
        self._pending_close = None

    @property
    def program_present(self) -> bool:
        return self._close_hit_ts is not None and self._last_ts - self._close_hit_ts < self._cfg.close_hold_s

    def promoted(self) -> list[Jingle]:
        """The jingles trusted right now (the hour of day can admit one a break early)."""
        return [j for j in self._jingles if self._trusted(j)]

    @property
    def voting(self) -> bool:
        """Silent until a sting has opened enough breaks to be trusted: a learner
        with nothing to say must not dilute the detectors that do (fusion norm)."""
        return any(j.kind == "open" for j in self.promoted())

    def vote(self, ts: float) -> DetectorVote:
        open_ts = self._open_hit_ts
        if open_ts is not None and ts - open_ts < self._cfg.hold_s and (
            self._close_hit_ts is None or self._close_hit_ts < open_ts
        ):
            family = f" family={self._last_variant}" if self._last_variant else ""
            return self._vote(
                ts, 1.0,
                f"jingle_open id={self._open_hit_id} agree={self._last_agreement:.2f} age_s={ts - open_ts:.1f}{family}",
            )
        known = len(self.promoted())
        return self._vote(ts, 0.0, f"jingle_quiet known={known} candidates={len(self._jingles) - known}")

    def describe(self) -> str:
        p = self.promoted()
        cands = len(self._jingles) - len(p)
        if not p:
            return f"learning ({cands} candidates; a sting must open 3 breaks)"
        opens = sum(1 for j in p if j.kind == "open")
        hours = sorted({h for j in p for h in j.hours})
        when = f"; heard at {', '.join(f'{h:02d}h' for h in hours)}" if hours else ""
        return f"{opens} opener(s), {len(p) - opens} closer(s) known, {cands} candidates{when}"

File: platforms/android/core/src/main/kotlin/io/adhush/core/Jingle.kt
package io.adhush.core

import java.io.File
import java.util.Calendar
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * Break jingles (ADR 0020, after AdVent): the short sting a channel plays
 * going into and out of every break. Instead of one fingerprint per
 * commercial, one per channel covers every break for good. Nothing is
 * hard-coded: every break the engine ends contributes the seconds around
 * its start (an opener) and its end (a closer) as *candidates*; a candidate
 * that turns out to open several different breaks is promoted to a jingle
 * and starts voting. An opener heard live ducks the set on its own (the
 * logo weight); a closer heard while ducked is programme evidence, which
 * ends a break as fast as a returning logo does.
 *
 * Families and hours (ADR 0027): a channel re-cuts its sting per show, so a
 * sting is also recognised transposed by a few semitones (the twelve chroma
 * bits rotate) or played a little faster or slower (the stored blocks
 * stretch), at a higher bar; a variant hit counts for the jingle it
 * resembles. Every jingle also remembers the local hours it opened breaks
 * in: heard again at one of those hours it is trusted a break earlier and
 * matched a little more loosely.
 */
enum class JingleKind(val wire: String) { OPEN("open"), CLOSE("close");
    companion object { fun of(s: String) = entries.firstOrNull { it.wire == s } ?: OPEN }
}

class Jingle(val id: Int, val kind: JingleKind, val blocks: List<Int>, var hits: Int, var falseHits: Int, val createdTs: Double,
             /** Local hours of day this sting opened (or closed) a break in. */
             val hours: MutableSet<Int> = HashSet())

/** The chroma word of the same block transposed up by [semitones] (down when negative): class c lives in bit 11 - c. */
fun pitchRotate(bits: Int, semitones: Int): Int {
    val k = ((semitones % Chroma.BITS) + Chroma.BITS) % Chroma.BITS
    val mask = (1 shl Chroma.BITS) - 1
    if (k == 0) return bits and mask
    return ((bits ushr k) or (bits shl (Chroma.BITS - k))) and mask
}

/** The same sting played at 1 / [factor] speed: 1.1 is ten per cent slower. */
fun tempoStretch(blocks: List<Int>, factor: Double): List<Int> {
    val n = max(1, (blocks.size * factor).roundToInt())
    return List(n) { i -> blocks[minOf(blocks.size - 1, (i / factor).toInt())] }
}

fun localHour(wall: Double): Int = Calendar.getInstance().apply { timeInMillis = (wall * 1000).toLong() }.get(Calendar.HOUR_OF_DAY)

interface JingleStore {
    fun load(): List<Jingle>
    fun save(jingles: List<Jingle>)
}

class FileJingleStore(private val file: File? = null) : JingleStore {
    override fun load(): List<Jingle> {
        val f = file?.takeIf { it.isFile } ?: return emptyList()
        val out = ArrayList<Jingle>()
        f.forEachLine { line ->
            val p = line.split('\t')
            if (p[0] == "j" && p.size >= 7) runCatching {
                val hours = if (p.size > 7) p[7].split(',').filter { it.isNotEmpty() }.map { it.toInt() }.toHashSet() else HashSet()  // v1 rows carry none
                out.add(Jingle(p[1].toInt(), JingleKind.of(p[2]), p[6].split(',').filter { it.isNotEmpty() }.map { it.toInt() }, p[3].toInt(), p[4].toInt(), p[5].toDouble(), hours))
            }
        }
        return out
    }
    override fun save(jingles: List<Jingle>) {
        val f = file ?: return
        f.parentFile?.mkdirs()
        val tmp = File(f.path + ".tmp")
        tmp.bufferedWriter().use { w ->
            w.write("# adhush jingles v2\tid\tkind\thits\tfalse_hits\tcreated\tblocks\thours\n")
            for (j in jingles) w.write("j\t${j.id}\t${j.kind.wire}\t${j.hits}\t${j.falseHits}\t${j.createdTs}\t${j.blocks.joinToString(",")}\t${j.hours.sorted().joinToString(",")}\n")
        }
        if (!tmp.renameTo(f)) { f.delete(); tmp.renameTo(f) }
    }
}

data class JingleConfig(
    val sampleIntervalS: Double = 0.5,
    /** AdVent measured three seconds as enough to recognise a sting. */
    val jingleS: Double = 3.0,
    /** A candidate keeps a wider window than the sting, so the sting can sit anywhere in it. */
    val candidateS: Double = 6.0,
    /** Aligned chroma-bit agreement that counts as the same sting (0.5 is chance). */
    val minAgreement: Double = 0.72,
    /** Distinct breaks a candidate must open (or close) before it votes. */
    val promoteHits: Int = 3,
    /** An opener's vote lasts this long: the first seconds of the break. */
    val holdS: Double = 20.0,
    /** A closer counts as programme evidence for this long. */
    val closeHoldS: Double = 10.0,
    /** Rolling audio kept, so a break's opener can still be cut out when the break ends. */
    val historyS: Double = 400.0,
    val maxCandidates: Int = 80,
    /** ADR 0027, families: a sting is also recognised transposed by up to this many semitones either way ... */
    val pitchShifts: Int = 2,
    /** ... and stretched by up to this fraction ... */
    val tempoTolerance: Double = 0.1,
    /** ... but a transposed or stretched match must clear a higher bar. */
    val familyPenalty: Double = 0.05,
    /** ADR 0027, hours: at an hour the sting has opened breaks in before, it is matched this much more loosely. */
    val hourBonus: Double = 0.03,
)

class JingleDetector(private val store: JingleStore, private val cfg: JingleConfig = JingleConfig()) : Detector {
    override val name = "jingle"
    private val ring = ArrayDeque<Pair<Double, Int>>()   // (ts, chroma bits) per sampleInterval
    private val pending = ArrayList<FloatArray>()
    private var pendingN = 0
    private var blockStart: Double? = null
    private val jingles = ArrayList<Jingle>()
    private var nextId = 1
    private var pendingClose: Double? = null
    private var openHitTs: Double? = null
    private var openHitId: Int? = null
    private var closeHitTs: Double? = null
    private var lastAgreement = 0.0
    private var lastVariant = ""
    private var lastTs = 0.0
    private var hour: Int? = null
    private var pendingCloseHour: Int? = null
    private val stingBlocks: Int get() = max(2, (cfg.jingleS / cfg.sampleIntervalS).toInt())
    private val candidateBlocks: Int get() = max(stingBlocks, (cfg.candidateS / cfg.sampleIntervalS).toInt())

    init { store.load().forEach { jingles.add(it); if (it.id >= nextId) nextId = it.id + 1 } }

    override fun warmup() { ring.clear(); pending.clear(); pendingN = 0; blockStart = null; pendingClose = null; openHitTs = null; closeHitTs = null }

    override fun observeAudio(block: AudioBlock) {
        var start = blockStart ?: block.ts
        pending.add(block.samples); pendingN += block.samples.size
        val need = Math.round(cfg.sampleIntervalS * block.sampleRate).toInt()
        while (pendingN >= need) {
            val all = FloatArray(pendingN); var o = 0
            for (p in pending) { System.arraycopy(p, 0, all, o, p.size); o += p.size }
            val head = all.copyOfRange(0, need); val rest = all.copyOfRange(need, all.size)
            ring.addLast(Pair(start, Chroma.chromaBits(head, block.sampleRate)))
            while (ring.size > (cfg.historyS / cfg.sampleIntervalS).toInt()) ring.removeFirst()
            start += need.toDouble() / block.sampleRate
            pending.clear(); if (rest.isNotEmpty()) pending.add(rest); pendingN = rest.size
            onSample(start)
        }
        blockStart = start
    }

    private fun blocksBetween(t0: Double, t1: Double): List<Int> = ring.filter { it.first >= t0 && it.first < t1 }.map { it.second }

    /** Best aligned agreement of the live sting against a candidate's wider window. */
    private fun bestAgreement(live: List<Int>, candidate: List<Int>): Double {
        if (live.size < stingBlocks || candidate.size < stingBlocks) return 0.0
        var best = 0.0
        for (off in 0..(candidate.size - live.size)) {
            val a = Chroma.agreement(live, candidate.subList(off, off + live.size))
            if (a > best) best = a
        }
        return best
    }

    /** The stored sting as heard, then every transposition and stretch the config allows, each labelled; the identity first and unlabelled. */
    private fun variants(blocks: List<Int>): List<Pair<List<Int>, String>> {
        val out = ArrayList<Pair<List<Int>, String>>()
        out.add(Pair(blocks, ""))
        val shifts = (-cfg.pitchShifts..cfg.pitchShifts).filter { it != 0 }
        val factors = if (cfg.tempoTolerance > 0) listOf(1.0 - cfg.tempoTolerance, 1.0 + cfg.tempoTolerance) else emptyList()
        for (k in listOf(0) + shifts) {
            val rotated = if (k != 0) blocks.map { pitchRotate(it, k) } else blocks
            for (f in factors) out.add(Pair(tempoStretch(rotated, f), "%+dst/%.2fx".format(k, f)))
            if (k != 0) out.add(Pair(rotated, "%+dst".format(k)))
        }
        return out
    }

    /** Best agreement of [live] against the sting or any variant of it, the variants docked [JingleConfig.familyPenalty] so the sting as heard wins a tie. */
    private fun bestFamily(live: List<Int>, jingleBlocks: List<Int>): Pair<Double, String> {
        var best = 0.0; var label = ""
        for ((blocks, variant) in variants(jingleBlocks)) {
            val a = bestAgreement(live, blocks) - (if (variant.isEmpty()) 0.0 else cfg.familyPenalty)
            if (a > best) { best = a; label = variant }
        }
        return Pair(best, label)
    }

    private fun sameSting(a: List<Int>, b: List<Int>): Boolean {
        // Slide the shorter core of one over the other: the sting may sit anywhere in either window.
        val core = stingBlocks
        for (i in 0..(a.size - core)) if (bestFamily(a.subList(i, i + core), b).first >= cfg.minAgreement) return true
        return false
    }

    /** Promoted outright, or one break short of it at an hour it has opened breaks in before. */
    private fun trusted(j: Jingle): Boolean {
        if (j.hits >= cfg.promoteHits) return true
        val h = hour ?: return false
        return h in j.hours && cfg.promoteHits - 1 > 0 && j.hits >= cfg.promoteHits - 1
    }

    private fun threshold(j: Jingle): Double = cfg.minAgreement - (if (hour != null && hour in j.hours) cfg.hourBonus else 0.0)

    private fun onSample(ts: Double) {
        lastTs = ts
        pendingClose?.let { end ->
            if (ts >= end + cfg.candidateS - 2.0) {
                pendingClose = null; val h = pendingCloseHour; pendingCloseHour = null
                addCandidate(JingleKind.CLOSE, blocksBetween(end - 2.0, end + cfg.candidateS - 2.0), ts, h)
            }
        }
        val live = ring.takeLast(stingBlocks).map { it.second }
        if (live.size < stingBlocks) return
        var best: Jingle? = null; var bestA = 0.0; var bestV = ""
        for (j in jingles) {
            if (!trusted(j)) continue
            val (a, variant) = bestFamily(live, j.blocks)
            if (a >= threshold(j) && a > bestA) { best = j; bestA = a; bestV = variant }
        }
        val j = best ?: return
        lastAgreement = bestA; lastVariant = bestV
        if (j.kind == JingleKind.OPEN) { if (openHitTs == null || ts - openHitTs!! > cfg.holdS) { openHitTs = ts; openHitId = j.id } }
        else closeHitTs = ts
    }

    private fun addCandidate(kind: JingleKind, blocks: List<Int>, ts: Double, hourOfDay: Int? = null) {
        if (blocks.size < stingBlocks) return
        val same = jingles.firstOrNull { it.kind == kind && sameSting(blocks, it.blocks) }
        if (same != null) { same.hits++; if (hourOfDay != null) same.hours.add(hourOfDay); store.save(jingles); return }
        jingles.add(Jingle(nextId++, kind, blocks, 1, 0, ts, if (hourOfDay != null) hashSetOf(hourOfDay) else HashSet()))
        // Keep the candidate list bounded: the oldest one-off candidates go first.
        while (jingles.count { it.hits < cfg.promoteHits } > cfg.maxCandidates) {
            val victim = jingles.filter { it.hits < cfg.promoteHits }.minByOrNull { it.createdTs } ?: break
            jingles.remove(victim)
        }
        store.save(jingles)
    }

    /**
     * A real break ran from [startTs] to [endTs] (media time): the seconds
     * around its start are an opener candidate now, the seconds around its
     * end a closer candidate once they have been heard.
     */
    fun learnBreak(startTs: Double, endTs: Double, wallStart: Double? = null) {
        val h = if (wallStart != null) localHour(wallStart) else hour
        addCandidate(JingleKind.OPEN, blocksBetween(startTs - (cfg.candidateS - 2.0), startTs + 2.0), endTs, h)
        pendingClose = endTs; pendingCloseHour = h
    }

    /** Wall-clock time passes: the hour of day steers which stings are trusted (ADR 0027). */
    fun tick(wall: Double) { hour = localHour(wall) }

    /** "Not an ad": a jingle that just fired was wrong; enough wrong calls demote it back to a candidate. */
    override fun userSaysProgramme(ts: Double) {
        val id = openHitId
        if (id != null && openHitTs != null && ts - openHitTs!! < 60.0) {
            jingles.firstOrNull { it.id == id }?.let { j -> j.falseHits++; if (j.falseHits >= j.hits) { j.hits = 0 }; store.save(jingles) }
        }
        openHitTs = null; openHitId = null; closeHitTs = null; pendingClose = null
    }

    /** A closer was just heard: the programme is back, whatever the audio methods think. */
    override val programPresent: Boolean get() = closeHitTs?.let { lastTs - it < cfg.closeHoldS } ?: false

    /** Silent until a sting has opened enough breaks to be trusted: a learner
     *  with nothing to say must not dilute the detectors that do (fusion norm). */
    override val voting: Boolean get() = promoted().any { it.kind == JingleKind.OPEN }

    override fun vote(ts: Double): DetectorVote {
        val open = openHitTs
        if (open != null && ts - open < cfg.holdS && (closeHitTs == null || closeHitTs!! < open)) {
            val family = if (lastVariant.isEmpty()) "" else " family=$lastVariant"
            return vote(ts, 1.0, "jingle_open id=$openHitId agree=${"%.2f".format(lastAgreement)} age_s=${"%.1f".format(ts - open)}$family")
        }
        return vote(ts, 0.0, "jingle_quiet known=${promoted().size} candidates=${jingles.size - promoted().size}")
    }

    /** The jingles trusted right now (the hour of day can admit one a break early). */
    fun promoted(): List<Jingle> = jingles.filter { trusted(it) }

    fun describe(): String {
        val p = promoted()
        val known = "${p.count { it.kind == JingleKind.OPEN }} opener${if (p.count { it.kind == JingleKind.OPEN } == 1) "" else "s"}, ${p.count { it.kind == JingleKind.CLOSE }} closer${if (p.count { it.kind == JingleKind.CLOSE } == 1) "" else "s"}"
        val cands = jingles.size - p.size
        val hours = p.flatMap { it.hours }.toSortedSet()
        val when_ = if (hours.isEmpty()) "" else "; heard at " + hours.joinToString(", ") { "%02dh".format(it) }
        return if (p.isEmpty()) "learning ($cands candidate${if (cands == 1) "" else "s"}; a sting must open 3 breaks)" else "$known known, $cands candidate${if (cands == 1) "" else "s"}$when_"
    }
}
